File: libs/power-log-parser/src/lib/parsers/minions-will-die-parser.ts
import { BlockType, GameTag, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider, GameEventHelper } from '../game-event';
import { Action, Node, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class MinionsWillDieParser implements ActionParser {
	readonly ParserName = 'MinionsWillDieParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.GameState &&
			node.Type === Action &&
			(node.Object as Action).Type === (BlockType.DEATHS as number)
		);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const action = node.Object as Action;
		const deathTags = action
			.GetDataRecursive()
			.filter((d): d is TagChange => d instanceof TagChange)
			.filter((t) => t.Name === (GameTag.ZONE as number) && t.Value === (Zone.GRAVEYARD as number))
			.filter((t) => this.GameState.CurrentEntities.get(t.Entity)?.IsMinionLike() ?? false);

		const deadMinions = deathTags.map((tag) => {
			const entity = this.GameState.CurrentEntities.get(tag.Entity)!;
			return {
				CardId: entity.CardId,
				EntityId: entity.Id,
				ControllerId: entity.GetEffectiveController(),
				Timestamp: tag.TimeStamp,
			};
		});

		return [
			GameEventProvider.Create(
				action.TimeStamp,
				'MINIONS_WILL_DIE',
				GameEventHelper.CreateProvider('MINIONS_WILL_DIE', null as any, -1, -1, this.StateFacade, {
					DeadMinions: deadMinions,
				}),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/MinionsWillDieParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Linq;

namespace HearthstoneReplays.Events.Parsers
{
    public class MinionsWillDieParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public MinionsWillDieParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.GameState
                && node.Type == typeof(Action)
                && (node.Object as Action).Type == (int)BlockType.DEATHS;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var action = node.Object as Action;
            var deathTags = action.GetDataRecursive()
                .Where(data => data is TagChange)
                .Select(data => data as TagChange)
                .Where(tag => tag.Name == (int)GameTag.ZONE && tag.Value == (int)Zone.GRAVEYARD)
                // Checking the current zone doesn't work, because we work on a Close node. However, since 
                // we are in the DEATHS block, we should be good
                //.Where(tag => GameState.CurrentEntities[tag.Entity].GetTag(GameTag.ZONE) == (int)Zone.PLAY)
                .Where(tag => GameState.CurrentEntities.GetValueOrDefault(tag.Entity)?.IsMinionLike() ?? false);
            var deadMinions = deathTags.Select(tag =>
            {
                var entity = GameState.CurrentEntities[tag.Entity];
                var cardId = entity.CardId;
                var controllerId = entity.GetEffectiveController();
                return new
                {
                    CardId = cardId,
                    EntityId = entity.Id,
                    ControllerId = controllerId,
                    Timestamp = tag.TimeStamp,
                };
            });

            //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, null, null);
            return new List<GameEventProvider> { GameEventProvider.Create(
                action.TimeStamp,
                "MINIONS_WILL_DIE",
                GameEvent.CreateProvider(
                    "MINIONS_WILL_DIE",
                    null,
                    -1,
                    -1,
                    StateFacade,
                    //gameState,
                    new {
                        DeadMinions = deadMinions,
                    }),
                true,
                node) };
        }
    }
}
